onchain/cairo/src/staking/staking.cairo

```diff
@@ -0,0 +1,312 @@
+        pub const TRANSFER_FAILED: felt252 = 'Transfer failed';
+    }
+
+    #[embeddable_as(StakingImpl)]
+    impl Staking<
+        TContractState, +HasComponent<TContractState>
+    > of IStaking<ComponentState<TContractState>> {
+        fn set_rewards_duration(ref self: ComponentState<TContractState>, duration: u256) {
+            let caller = get_caller_address();
+            assert(caller == self.owner.read(), Errors::NOT_AUTHORIZED);
+
+            let block_timestamp: u256 = get_block_timestamp().try_into().unwrap();
+            assert(self.finish_at.read() < block_timestamp, Errors::REWARD_DURATION_NOT_FINISHED);
+
+            self.duration.write(duration);
+        }
+
+        fn notify_reward_amount(ref self: ComponentState<TContractState>, amount: u256) {
+            let caller = get_caller_address();
+            let this_contract = get_contract_address();
+
+            assert(caller == self.owner.read(), Errors::NOT_AUTHORIZED);
+
+            let zero_address = self.zero_address();
+
+            self.update_reward(zero_address);
+
+            let block_timestamp: u256 = get_block_timestamp().try_into().unwrap();
+
+            let rewards_token = IERC20Dispatcher { contract_address: self.rewards_token.read() };
+
+            let transfer_from = rewards_token.transfer_from(caller, this_contract, amount);
+            assert(transfer_from, Errors::TRANSFER_FAILED);
+
+            if block_timestamp >= self.finish_at.read() {
+                self.reward_rate.write(amount / self.duration.read())
+            } else {
+                let remaining_rewards = (self.finish_at.read() - block_timestamp)
+                    * self.reward_rate.read();
+
+                self.reward_rate.write((amount + remaining_rewards) / self.duration.read());
+            }
+
+            assert(self.reward_rate.read() > 0, Errors::REWARD_RATE_IS_ZERO);
+            assert(
+                self.reward_rate.read()
+                    * self.duration.read() <= rewards_token.balance_of(this_contract),
+                Errors::REWARD_AMOUNT_GREATER_THAN_CONTRACT_BALANCE
+            );
+
+            self.finish_at.write(get_block_timestamp().try_into().unwrap() + self.duration.read());
+            self.updated_at.write(get_block_timestamp().try_into().unwrap());
+        }
+
+        fn stake(ref self: ComponentState<TContractState>, amount: u256) {
+            let caller = get_caller_address();
+            let this_contract = get_contract_address();
+
+            self.update_reward(caller);
+
+            assert(amount > 0, Errors::AMOUNT_IS_ZERO);
+            let staking_token = IERC20Dispatcher { contract_address: self.staking_token.read() };
+            let transfer = staking_token.transfer_from(caller, this_contract, amount);
+
+            assert(transfer, Errors::TRANSFER_FAILED);
+
+            let prev_stake = self.balance_of.entry(caller).read();
```

Comment:
A fast deposit stake can broke the reward thing, because we need to check which deposit with a specific duration (having a list basically).

Otherwise fast/withdrawn deposit and multi deposit broke the amount to be rewarded.

Reply:
I don't think so, a user can stake multiple times. It's doesn't break. This implementation is the exact same implementation of the Synthetix Staking Rewards contract: https://github.com/Synthetixio/synthetix/blob/develop/contracts/StakingRewards.sol